kf_lib_data_ingest/common/misc.py

```diff
@@ -128,3 +132,163 @@ def multisplit(string: str, delimiters: list):
+    :param read: total retries on read errors
+    :param connect: total retries on connection errors
+    :param status: total retries on bad status codes defined in
+    `status_forcelist`
+    :param backoff_factor: affects sleep time between retries
+    :param status_forcelist: list of HTTP status codes that force retry
+    """
+    session = session or requests.Session()
+
+    retry = Retry(
+        total=total,
+        read=read,
+        connect=connect,
+        backoff_factor=backoff_factor,
+        status_forcelist=status_forcelist,
+        method_whitelist=False
+    )
+    adapter = HTTPAdapter(max_retries=retry)
+    session.mount('http://', adapter)
+    session.mount('https://', adapter)
+
+    return session
+
+
+def get_swagger_schema(url, entity_names,
+                       cached_schema_filepath=None, logger=None):
+    """
+    Get schemas for entities in the target API using the API's swagger endpoint
+
+    Will extract parts of the swagger response to create the output dict
+
+    Example swagger response:
+    {
+        'info': {
+            'version': '1.9.0',
+            'description':  'stuff',
+            ...
+        },
+        'definitions': {
+            'Participant': {
+                ...
+            },
+            'Biospecimen': {
+                ...
+            },
+            'BiospecimenResponse': {
+                ...
+            }
+            ...
+        }
+    }
+
+    Will turn into the output:
+
+    {
+        'target_service': https://kf-api-dataservice.kidsfirstdrc.org,
+        'version': 1.9.0,
+        'definitions': {
+            'participant': {
+                ...
+            },
+            'biospecimen': {
+                ...
+            },
+            ...
+        }
+    }
+
+    Items in entity_names must be snake cased versions of existing keys in
+    swagger 'definitions'.
+
+    See https://petstore.swagger.io/v2/swagger.json for example
+    definitions content/structure
+
+    :param url: URL to a target service that implements a Swagger API and
+    has a /swagger endpoint
+    :param entity_names: list of snake cased names of entities to extract from
+    swagger definitions
+    :param cached_schema_filepath: file path to a JSON file containing a
+    saved version of the target service's schema.
+    :param logger: logger to use when reporting errors
+    :returns output: a dict with the schema definition and version
+    """
+    output = None
+    err = None
+    common_msg = 'Unable to retrieve target schema from target service!'
+    schema_url = f'{url}/swagger'
+
+    # Try to get schemas and version from the target service
+    try:
+
+        response = requests_retry_session().get(schema_url)
+
+    except ConnectionError as e:
+
+        err = f'{common_msg}\nCaused by {str(e)}'
+
+    else:
+        if response.status_code == 200:
+            # API Version
+            version = response.json()['info']['version']
```

Comment:
specs can be yaml also https://github.com/OAI/OpenAPI-Specification/blob/master/examples/v2.0/yaml/petstore-with-external-docs.yaml

Doesn't need to be done here, but we should add an issue to do that later if not.

Reply:
👍